### PR TITLE
Instance rendering in gloo.

### PR DESCRIPTION
**Pull request by brisvag**
Following up on #2376, I took a deep dive into the glir and gloo code and found a way to enable a reasonable high level interface for instanced rendering. I'm sure there's a bunch of ways to do this, but the way that made the most sense to me is this one. To summarize the changes:

- `VertexBuffer` now takes a `divisor` argument, which marks the vertex buffer as instance-based (might be worth renaming this at the highest level at least)
- when the above vertex buffer gets assigned as an `attribute` to a program, the divisor value will be forwarded to `glir`
- when `draw()` gets called from the program, on top of checking that normal attributes have the same size, it will also check that for each attribute with a divisor, `n_instances = size * divisor` is the same. A `DRAW` command is then sent to `glir`, with `instances=n_instances` if there were any buffers with divisors, otherwise with `instances=1`

on the `glir` side:
- before draw, `_pre_draw()` does all the binding/enabling of buffers. Here, buffers with a `divisor` will get the `bl.glVertexAttribDivisor` set appropriately
- `GlirProgram.draw()` will then call the appropriate draw function `instanced` if `instances > 1`, otherwise the normal one.

I added a couple examples based on existing ones which should exemplify the behaviour. Everything works also with `Visuals` and their multiprogram machinery!

Here's the new colored cube example, which instances 100 times the same cube with 100 different shifts and 5 different colors:

https://user-images.githubusercontent.com/23482191/190858605-68574f32-5c13-4170-9333-b2cc9e2ff943.mp4

### CODE REVIEW

**Review by djhoese — COMMENTED** [review]
I definitely don't know enough to give this a proper review, but I can run some ideas by you to see what you thought.

1. Could `divisor` be a keyword argument with a default of `0` for backwards compatibility?
2. Could/should this functionality be implemented as a subclass of VertexBuffer? The changes seem simple enough to not need it, but I wanted to bring it up.
3. I would prefer if any new examples, and any examples they reference, be placed in the newer gallery `examples/gloo`, `examples/scene`, or `examples/plotting` directories so they show up in the website gallery.

**Comment by djhoese** [comment]
@vispy/core Please review when you have the time.

**Comment by brisvag** [comment]
> Could `divisor` be a keyword argument with a default of `0` for backwards compatibility?

Yes, good point... I did make it a "silent" keyword in a couple of places, but I should probably do it everywhere.

> Could/should this functionality be implemented as a subclass of VertexBuffer? The changes seem simple enough to not need it, but I wanted to bring it up.

I thought about that... I'm not sure if it's worth the "overhead" of an extra class, the bulk of the logic happens at lower levels anyways. Happy to change if that seems like a better solution though, I don't feel strongly about this.

> I would prefer if any new examples, and any examples they reference, be placed in the newer gallery `examples/gloo`, `examples/scene`, or `examples/plotting` directories so they show up in the website gallery.

Sure! I didn't know about these distinctions. Is it just those 3 that get generated?

**Comment by djhoese** [comment]
Regarding the gallery directories, yeah it was something I didn't know was a restriction until I added the gallery. `sphinx-gallery` works on single directories with only one level of sub-directories. To get the structure we needed I split it into 3 separate "galleries" (gloo, scene, plotting). Multiple galleries is something supported by sphinx-gallery.

https://vispy.org/gallery/

Regarding the class, my thought was more on the user or developer cognitive load. Like, if they are reading the main VertexBuffer class then transitioning to a specialized instance-based class would be easier than trying to understand it all in one class. That said, it probably isn't a big deal and a single class is probably fine.

**Comment by brisvag** [comment]
I moved the examples to `gloo` and `scene` (not sure if the latter makes sense?).

**Comment by djhoese** [comment]
Yeah that `scene` example is a little "light" but maybe that's OK. I would say move it to gloo, but I purposefully didn't make a gallery between `gloo` and `scene` because in my opinion not many people need Visuals without a SceneCanvas. They work fine with a regular Canvas, but we have to draw the line in the examples at some point.

Thanks for moving the examples. I think now we just have to wait for what others think.

File: vispy/gloo/buffer.py [hunk 1]
```diff
@@ -449,6 +453,9 @@ def _prepare_data(self, data, convert=False):
             self._last_dim = c
         return data
 
+    def set_instancing_divisor(self, divisor):
```
Comment (djhoese) [inline vispy/gloo/buffer.py:456]: Does this need to be a public method? If not, does it need to be a method at all?

**Comment by brisvag** [comment]
TODO: update GLIR docstring

File: vispy/gloo/buffer.py (hunk 1 above)
Comment (brisvag) [reply vispy/gloo/buffer.py:456]: I figured it would be nice to expose a way to change this afterwards, but It's not strictly needed.

Comment (brisvag) [reply vispy/gloo/buffer.py:456]: Yeah, good point!

**Review by djhoese — APPROVED** [review]
I approve of this as it is, but wouldn't mind if `divisor` and `instances` in the `glir.py` were turned into keyword arguments with defaults to make a single instance.

**Comment by brisvag** [comment]
Updated the branch to main, and made the arguments into keywords. `GLIR` commands will still require the arguments, but I think it's better to leave as is rather than introducing optional arguments in GLIR for the first time.

**Comment by djhoese** [comment]
> it's better to leave as is rather than introducing optional arguments in GLIR for the first time.

Agreed.

**Comment by brisvag** [comment]
As agreed in the monthly meeting (and since no negative reviews came in since then) I'm merging this :)

**Pull request merged**